README.md
## Data Rapid

Data Rapid is a data generation tool integrated with R  for creating real frequency distribution data as well as constant data.Its a web based test data generation tool which uses regex pattern and predefined static content to generate the realistic data independent of any domain. Data Rapid eliminates the need of coding to generate the data. Data Rapid adds the capabilities of constant data generation and distribution data generation. With rapidly changing dynamics in market place, extracting insights really fast to support real-time business processes have become a must-have competitive advantage. Data Rapid enables enterprises to build meaningful constant and real data there by building the correlations across diverse data sets.
    
<img src="https://raw.githubusercontent.com/Infosys/data-rapid/master/landing.PNG">     

**Generate realistic huge data-sets within seconds**

Why Data Rapid? – We all have been working on different POC&#39;s (Proof of concept) and for every scenario we require huge data sets. In many cases we do not get the data from the customer and then we have to generate the data ourselves. In such a case it becomes difficult to create data manually. In such situations the data rapid comes in handy which enables us to generate huge datasets in a very short span of time.

 
**Data Rapid Builder Details**

The Data Rapid can be used to generate data for any of the simple as well as complex constant and real distribution data.


Some of the features of this tool are as follows

- A template based data generation which has a visual code free development
- Create large volumes of data within a couple of clicks.
- Generate meaningful test data at row level.
- Extremely fast data generation (Generates more than 10 Million records in less than 3 minutes).
- Improve operational efficiency.
- Generates the statistical distribution data as well as constant data.
- Facility to group the column values into factors/segments using column bucketing functionality.
- Generate user defined data using the regex pattern.
- Generate huge log files within seconds.
- Facility to download the file directly to your system.
- Reuse the configuration after saving the dataset.
- Sftp/ftp transfer facility for the files.

**Builder Data Types**

The Data Builder feature allows us to build constant data such as country names, zip-codes, phone numbers, and alpha-numeric codes and so on. The user can generate any specific data which he wants using the simple java regex pattern. If there are use cases wherein he wants to specify certain repetitive values, he can generate that data also.

The supported data types are AlphaNumeric, AmericanCard, City, Country, CVV, Date, Default Set, Digit Format, EmailAddress, FirstName, FloatRange, Guid, HexadecimalCode,  HexColors,  IntegerRange, IPAddress, LastName, Location, MACAddress, MastercardNumber, NumberFormat, Discovery Card number, Password, PhoneNumber, PhoneNumberWithExt, RCommand, SSN, State, Timestamp, UniqueValues, UserDefiendRegexPattern, VisaCreditCardNumber, ZipCode, IMEI number, DiscoveryCreditCard and  IncrementalUniqueValues.

**Available Data Types**

The following are the data builder type configuration details.

**Data Rapid Data Types**

| **Sr No.** | **Data Types** | **Comments** | **Values  field Samples** |
| --- | --- | --- | --- |
| 1 | Alpha Numeric  | Generate Alphanumeric codes based on the parameters in values field. Usage : Select drop down Alpha Numeric and set the values field in the format \&lt;numberOfDigits,numberOfAlphabets\&gt; | Eg : 3,2 Will generates an alpha numeric code like 459db |
| 2 | American Card  | Generate Americancard numbers automatically | No values allowed |
| 3 | City  | Generate City automatically | No values allowed |
| 4 | Country  | Generate Country automatically | No values allowed |
| 5 | CVV  | Generate CVV numbers automatically | No values allowed |
| 6 | Date  | Generate random dates between start date and the end date based on the parameters in values field. Usage : Select drop down Date and set the values field in the format \&lt;start date in dd/MM/yyyy\&gt;-\&lt;end date in dd/MM/yyyy\&gt;#\&lt;output date format\&gt; | Eg : 01/01/2014-01/01/2016#dd/MMM/yyyy  will generate dates between 01/01/2014 and 01/01/2016 in dd/MMM/yyyy format |
| 7 | Default Set  | Generate user defined values from the set provided in the values field separated by comma. Usage : Select drop down Default Set and set the values field in the format \&lt;attributes separated by comma\&gt; | Eg: Machine1,Machine2,Machine3  Will generate values from any of the Machine1 or Machine2 or Machine3 |
| 8 | Digit format  | Generate random digits based on the parameters in values field.  Usage : Select drop down Digit Format and set the values field in the format \&lt;n digit number\&gt; | Eg: 5  Will generate numbers like 45689 |
| 9 | Discovery Credit Card  | Generate DiscoveryCreditCard numbers automatically | No values allowed |
| 10 | Email Address  | Generate emailaddress automatically | No values allowed |
| 11 | First Name  | Generate firstname automatically | No values allowed |
| 13 | Float Range  | Generate random float values between the given ranges. Usage : \&lt;lower range\&gt; to \&lt;upper range\&gt; | Eg : 1 to 10000 Will generate float range between 1 and 10000 |
| 14 | Guid(Globally Unique Identifier)  | Generate guid automatically | No values allowed |
| 15 | Hexa Decimal Code  | Generate hexadecimalcode automatically | No values allowed |
| 16 | Hex Colors  | Generate hexcolors automatically | No values allowed |
| 17 | IncrementalUnique Values  | Generate IncrementalUniqueValues between the range. Specify the start and end value to generate the unique key and the value by which you want to increment the values. Usage : \&lt;lower range\&gt;to\&lt;Upper Range\&gt;,\&lt;increment sequence\&gt; | Eg : 10 to 100,2 Will generate the values between 10 and 100 increment by 2 |
| 18 | Integer Range  | Generate random integer values between the given ranges.Usage : \&lt;lower range\&gt;to\&lt;Upper Range\&gt; | Eg : -100 to 100 Will generate numbers between -100 to 100 |
| 19 | IMEI Number(International Mobile Station Equipment Identity)  | Generate imei number automatically | No values allowed |
| 20 | IP Address  | Generate ip-address automatically | No values allowed |
| 21 | Last Name  | Generate last name automatically | No values allowed |
| 22 | Location  | Generate location automatically | No values allowed |
| 23 | MAC Address  | Generate macaddress automatically | No values allowed |
| 24 | MastercardNumber  | Generate mastercardnumber automatically | No values allowed |
| 25 | Number Format  | Generate numbers n,m with length m and precision of m. | Eg: 4,2  Will generate numbers like 45.23 |
| 26 | Password  | Generate password automatically | No values allowed |
| 27 | Phone Number  | Generate phonenumber automatically | No values allowed |
| 28 | Phone Number WithExt  | Generate phonenumberwithext automatically | No values allowed |
| 29 | RCommand  | Can be used to provide the distribution R command. Execute the rcommand specified. The value of n should be equal to the total number of rows.R Enabled distribution data generated will reverse the seed generation with respect to the expected frequency distribution like rnorm(4, mean=50, sd=10), floor(runif(100, min=0, max=101)) etc.  | Eg: rnorm(n=1000,mean=100,sd=5.2) Will generate distribution values with mean 100 and standard deviation 5.2 |
| 30 | SSN  | Generate ssn automatically | No values allowed |
| 31 | State  | Generate state automatically | No values allowed |
| 32 | Timestamp  | Generate the timestamp.Enter the fromDate-ToDate to generate standard Timestamp. | Eg: 01/01/2014-01/01/2016 Will generate timestamp values between 01/01/2014 and 01/01/2016 |
| 33 | Unique Values  | Generate uniquevalues. Specify the start value and end value to generate the unique key. | Eg : 1 to 1000 Will generate unique values between 1 and 1000 |
| 34 | User DefinedRegex Pattern  | Generate user defined regex pattern. |  Eg : [0-9]{2} Will generate a number with the regex pattern |
| 35 | VisaCreditCard Number  | Generate visacreditcardnumber automatically | No values allowed |
| 36 | ZipCode | Generate zipcode automatically | No values allowed |

**Sample Data Generation Details**

Following are the sample data generation configurations.

**Customer Data Set**

| **Sr No.** | **Column Header** | **Data Type** | **Values** | **Comments** |
| --- | --- | --- | --- | --- |
| 1 | DateId | Unique Values |  1 to 1000 | Will generate the unique values from 1 till 1000 |
| 2 | Date | Date |  01/01/2014-01/01/2016#dd/MMM/yyyy | Will generate the dates between 1st Jan 2014 to 1st Jan 2016 and will generate the output in the user defined output format  |
| 3 | CustomerId | Unique Values |  1 to 1000 | Will generate the unique values from 1 till 1000 |
| 4 | Customer Name | First Names |   | Automatically Populated |
| 5 | Customer Location | Location |   | Automatically Populated |
| 6 | Zip Code | Zip Code |   | Will generate the zip codes. |
| 7 | RequestChannelId | Integer Range | 1 to 4 | Will generate the numbers from 1 to 4 randomly |
| 8 | RequestChannel | Default Set | Call,Email,Online,Facebook | Will generate the data from the column values that we have given. |
| 9 | TranChannelId | Integer Range | 1 to 3 | Will generate the numbers from 1 to 3 randomly |
| 10 | TranChannel | Default Set | Store,Mobile App,Web | Will generate the data from the column values that we have given. |
| 11 | ProblemAreaId | Integer Range | 1 to 25 | Will generate the numbers from 1 to 25 randomly |
| 12 | ProblemArea | Default Set | Shopping,Order,Payment,Wallet,Others,Shipping | Will generate the data from the column values that we have given. |
| 13 | ProblemType | Default Set | Ambiguios product description,Incorrect delivery,Taxes,Unable to fund,Image not available,Shipping charges | Will generate the data from the column values that we have given. |
| 14 | NumberofInteractions | Integer Range | 1 to 3 | Will generate the numbers from 1 to 3 randomly |

**Log File Data Set**

| **Sr No.** | **Column Header** | **Data Type** | **Values** | **Comments** |
| --- | --- | --- | --- | --- |
| 1 | Date | Date | 01/01/2014-01/01/2016#dd/MMM/yyyy | Will generate the dates between 1st Jan 2014 to 31 Dec 2014 |
| 2 | Time | Date | 01/01/2014-01/01/2016#HH:mm:ss | Will generate the time |
| 3 | cs-ip | IpAddress |   | Will generate the ip addresses |
| 4 | sc-bytes | Integer Range | 1 to 255   | Will generate the numbers from 1 to 255 randomly |
| 5 | cs-method | Default Set |  GET,POST | Will generate method according to the values given |
| 6 | Phone number | PhoneNumber |   | Will generate the phone numbers. |


RELEASE-NOTES
#
# Copyright 2018 Infosys Ltd.
#
# Licensed to the Apache Software Foundation (ASF) under one or more
# contributor license agreements.  See the NOTICE file distributed with
# this work for additional information regarding copyright ownership.
# The ASF licenses this file to You under the Apache License, Version 2.0
# (the "License"); you may not use this file except in compliance with
# the License.  You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

This is not an official release notes document.  It exists for developers
to jot down their notes while working in the source code.

###########################################################
# 1.1.0
###########################################################

Existing features
--------------------------------
- Listing the currently available features
     Create large volumes of data within a couple of clicks.
     Generate meaningful test data at row level.
     Extremely fast data generation (Generates more than 10 Million records in less than 3 minutes).
     Improve operational efficiency.
     Generates the real time data (statistical distributions) as well as constant data.
     Facility to group the column values into factors/segments using column bucketing functionality.
     Generate user defined data using the regex pattern.
     Generate huge log files within seconds.
     Facility to download the file directly to your system.
     Reuse the configuration after saving the dataset.
     Sftp/ftp transfer facility for the files.



datarapid-core/src/main/java/org/datarapid/core/StarterClass.java
/*
 * Copyright 2018 Infosys Ltd.
 *
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.datarapid.core;

import org.datarapid.core.services.DataGenCoreServiceImpl;
import org.datarapid.core.view.ColumnMeta;
import org.datarapid.core.view.FileInfo;
import org.datarapid.core.view.JobConfiguration;
import org.springframework.context.ApplicationContext;
import org.springframework.context.support.ClassPathXmlApplicationContext;

import java.util.ArrayList;
import java.util.List;

/**
 * @Description : Starter Class
 */
public class StarterClass {

    public static void main(String[] args) {

        //ColumnMapper columnMapper1 = new ColumnMapper("Aaron", "cnst","expression", "defaultset", "50","");

        ColumnMeta meta1 = new ColumnMeta("firstname", "cnst", "expression", "FirstName", "Aaron,Arun", "1", null);
        ColumnMeta meta2 = new ColumnMeta("LastNames", "cnst", "expression", "lastname", "", "2", null);

        ColumnMeta meta3 = new ColumnMeta("Location", "cnst", "expression", "location", "", "3", null);
        ColumnMeta meta4 = new ColumnMeta("IpAddress", "cnst", "expression", "IpAddress", "", "4", null);
        ColumnMeta meta5 = new ColumnMeta("EmailAddresses", "cnst", "expression", "emailaddress", "", "5", null);
        ColumnMeta meta6 = new ColumnMeta("Problem Areas", "cnst", "expression", "defaultset", "A,B,C,D,E,F", "6", null);

        ColumnMeta meta7 = new ColumnMeta("IntegerRange", "cnst", "expression", "IntegerRange", "-1000to500", "7",
                null);
        ColumnMeta meta8 = new ColumnMeta("NumberFormat", "cnst", "expression", "NumberFormat", "4,2", "8", null);
        ColumnMeta meta9 = new ColumnMeta("Digits", "cnst", "expression", "digitFormat", "6", "9", null);
        ColumnMeta meta10 = new ColumnMeta("PhoneNumber", "cnst", "expression", "phonenumber", "", "10", null);

        ColumnMeta meta11 = new ColumnMeta("Date", "cnst", "expression", "Date", "01/01/2010-01/01/2015#dd/MMM/yyyy",
                "11", null);

        ColumnMeta meta12 = new ColumnMeta("TimeStamp", "cnst", "expression", "TimeStamp", "01/01/2010-01/01/2015",
                "12", null);

        ColumnMeta meta13 = new ColumnMeta("alphanumeric", "cnst", "expression", "alphanumeric", "2,2", "13", null);

        ColumnMeta meta14 = new ColumnMeta("city", "cnst", "expression", "city", "", "14", null);
        ColumnMeta meta15 = new ColumnMeta("Zipcode", "cnst", "expression", "zipcode", "", "15", null);
        ColumnMeta meta16 = new ColumnMeta("country", "cnst", "expression", "country", "", "16", null);

        ColumnMeta meta17 = new ColumnMeta("HexadecimalCode", "cnst", "expression", "HexadecimalCode", "", "17", null);

        ColumnMeta meta18 = new ColumnMeta("Password", "cnst", "expression", "Password", "", "18", null);
        ColumnMeta meta19 = new ColumnMeta("visaCreditCardNumber", "cnst", "expression", "visaCreditCardNumber", "",
                "19", null);
        ColumnMeta meta20 = new ColumnMeta("mastercardNumber", "cnst", "expression", "mastercardNumber", "", "20",
                null);
        ColumnMeta meta21 = new ColumnMeta("AmericanCard", "cnst", "expression", "AmericanCard", "", "21", null);
        ColumnMeta meta22 = new ColumnMeta("PanCardNumber", "cnst", "expression", "PanCardNumber", "", "22", null);
        ColumnMeta meta23 = new ColumnMeta("cvv", "cnst", "expression", "cvv", "", "23", null);
        ColumnMeta meta24 = new ColumnMeta("hexColors", "cnst", "expression", "hexColors", "", "24", null);
        ColumnMeta meta25 = new ColumnMeta("ssn", "cnst", "expression", "ssn", "", "25", null);
        ColumnMeta meta26 = new ColumnMeta("PhoneNumberWithExt", "cnst", "expression", "PhoneNumberWithExt", "", "26",
                null);

        ColumnMeta meta27 = new ColumnMeta("guid", "cnst", "expression", "guid", "", "27", null);

        ColumnMeta meta28 = new ColumnMeta("UniqueValues", "cnst", "expression", "UniqueValues", "1-1000", "28", null);
        ColumnMeta meta29 = new ColumnMeta("Rcolumn1", "cnst", "expression", "RCommand",
                "rnorm(n=100000,mean=105,sd=5.2)", "29", null);
        ColumnMeta meta30 = new ColumnMeta("User defined Pattern", "cnst", "expression", "UserDefinedRegex",
                "[1-9]{3}\\.[0-9]{2}", "30", null);

        ColumnMeta meta31 = new ColumnMeta("MACaddress", "cnst", "expression", "MACaddress", "", "31", null);

        ColumnMeta meta32 = new ColumnMeta("FloatRange", "cnst", "expression", "FLOATRANGE", "-250 to 1000", "32",
                null);

        // *************************************************************************************************************************************//

        // for bucketing purpose only

		/*
         * ColumnMapper columnMapper1 = new ColumnMapper("a", "cnst",
		 * "expression", "Rcommand", "10","rnorm(n=10000,mean=101,sd=5.2)");
		 *
		 * ColumnMapper columnMapper2 = new ColumnMapper("b", "cnst",
		 * "expression", "NumberFormat", "10", "4,2"); ColumnMapper
		 * columnMapper3 = new ColumnMapper("c", "cnst", "expression",
		 * "DigitFormat", "10", "4"); ColumnMapper columnMapper4 = new
		 * ColumnMapper("d", "cnst", "expression", "Ipaddress", "10", "4,2");
		 * ColumnMapper columnMapper5 = new ColumnMapper("e", "cnst",
		 * "expression", "PanCardNumber", "10", ""); ColumnMapper columnMapper6
		 * = new ColumnMapper("f", "cnst", "expression", "mastercardNumber",
		 * "10", ""); ColumnMapper columnMapper7 = new ColumnMapper("g", "cnst",
		 * "expression", "HexColors", "10", ""); ColumnMapper columnMapper8 =
		 * new ColumnMapper("h", "cnst", "expression", "FirstName", "10", "");
		 * ColumnMapper columnMapper9 = new ColumnMapper("i", "cnst",
		 * "expression", "EmailAddress", "10", ""); ColumnMapper columnMapper10
		 * = new ColumnMapper("j", "cnst", "expression", "VisaCreditCardNumber",
		 * "10", "");
		 *
		 *
		 * // ColumnMapper columnMapper3=new //
		 * ColumnMapper(valuesFromSet,type,valueType, //
		 * dataType,percentageOfrows,values)
		 *
		 * // ColumnMapper columnMapper3 = new ColumnMapper("CCC", "cnst", //
		 * "expression", "defaultSet", "a,b,c,d,e", "");
		 *
		 * List<ColumnMapper> listMapper = new ArrayList<ColumnMapper>();
		 * listMapper.add(columnMapper1); listMapper.add(columnMapper2);
		 * listMapper.add(columnMapper3); listMapper.add(columnMapper4);
		 * listMapper.add(columnMapper5); listMapper.add(columnMapper6);
		 * listMapper.add(columnMapper7); listMapper.add(columnMapper8);
		 * listMapper.add(columnMapper9); listMapper.add(columnMapper10);
		 *
		 * BucketedColumn bucketedColumn = new
		 * BucketedColumn("BucketedColumnName", listMapper); ColumnMeta
		 * metaBucket = new ColumnMeta("ColumnParent", "cnst", "EXPRESION",
		 * "firstName", "", "1", bucketedColumn);
		 *
		 * ColumnMeta metaBucket2 = new ColumnMeta("ColumnParent", "cnst",
		 * "EXPRESION", "firstName", "", "1",bucketedColumn);
		 */

        // ******************************************************************************************************************

        List<ColumnMeta> list = new ArrayList<ColumnMeta>();
		/*
		 * list.add(metaBucket2); list.add(metaBucket);
		 */
        list.add(meta1);
        list.add(meta2);
        list.add(meta3);
        list.add(meta4);
        list.add(meta5);
        list.add(meta6);
        list.add(meta7);
        list.add(meta8);
        list.add(meta9);
        list.add(meta10);
        list.add(meta11);
        list.add(meta12);
        list.add(meta13);
        list.add(meta14);
        list.add(meta15);
        list.add(meta16);
        list.add(meta17);
        list.add(meta18);
        list.add(meta19);
        list.add(meta20);
        list.add(meta21);
        list.add(meta22);
        list.add(meta23);
        list.add(meta24);
        list.add(meta25);
        list.add(meta26);
        list.add(meta27);
        list.add(meta28);
        list.add(meta29);
        list.add(meta30);
        list.add(meta31);
        list.add(meta32);

        FileInfo fileInfo = new FileInfo(list);
        JobConfiguration configuration = new JobConfiguration("testingFile.csv", fileInfo, "1000", "CSV");

        @SuppressWarnings("resource")
        ApplicationContext context = new ClassPathXmlApplicationContext("datagen_core.xml");
        DataGenCoreServiceImpl dataGenCoreServiceImpl = (DataGenCoreServiceImpl) context
                .getBean("dataGenCoreServiceImpl");

        try {

            String finalMessage = dataGenCoreServiceImpl.buildData(configuration);
            System.out.println(finalMessage);
        } catch (Exception e) {
            System.out.println("in exception");
            e.printStackTrace();
        }

        System.out.println("over and out");
    }
}
